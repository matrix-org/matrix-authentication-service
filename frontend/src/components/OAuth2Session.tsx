// Copyright 2022 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Button } from "@vector-im/compound-web";
import { atom, useSetAtom } from "jotai";
import { atomFamily } from "jotai/utils";
import { atomWithMutation } from "jotai-urql";
import { useTransition } from "react";

import { FragmentType, graphql, useFragment } from "../gql";
import { getDeviceIdFromScope } from "../utils/deviceIdFromScope";
// import LoadingSpinner from "./LoadingSpinner/LoadingSpinner";
import { Session } from "./Session/Session";

const FRAGMENT = graphql(/* GraphQL */ `
  fragment OAuth2Session_session on Oauth2Session {
    id
    scope
    createdAt
    finishedAt
    client {
      id
      clientId
      clientName
      clientUri
    }
  }
`);

type Oauth2SessionType = {
  id: string;
  scope: string;
  createdAt: number;
  finishedAt?: number;
  client: {
    id: string;
    clientId: string;
    clientName: string;
    clientUri: string;
  }
}

const END_SESSION_MUTATION = graphql(/* GraphQL */ `
  mutation EndOAuth2Session($id: ID!) {
    endOauth2Session(input: { oauth2SessionId: $id }) {
      status
      oauth2Session {
        id
        ...OAuth2Session_session
      }
    }
  }
`);

const endSessionFamily = atomFamily((id: string) => {
  const endSession = atomWithMutation(END_SESSION_MUTATION);

  // A proxy atom which pre-sets the id variable in the mutation
  const endSessionAtom = atom(
    (get) => get(endSession),
    (get, set) => set(endSession, { id }),
  );

  return endSessionAtom;
});

type Props = {
  session: FragmentType<typeof FRAGMENT>;
};

const OAuth2Session: React.FC<Props> = ({ session }) => {
  const [pending, startTransition] = useTransition();
  const data = useFragment<Oauth2SessionType>(FRAGMENT, session);
  const endSession = useSetAtom(endSessionFamily(data.id));

  // @TODO(kerrya) make this wait for session refresh properly
  // https://github.com/matrix-org/matrix-authentication-service/issues/1533
  const onSessionEnd = (): void => {
    startTransition(() => {
      endSession();
    });
  };

  const sessionName = getDeviceIdFromScope(data.scope);

  return (
    <Session
      id={data.id}
      name={sessionName}
      createdAt={data.createdAt}
      finishedAt={data.finishedAt}
      clientName={data.client.clientName}
    >
      {!data.finishedAt && (
        <Button
          kind="destructive"
          size="sm"
          onClick={onSessionEnd}
          disabled={pending}
        >
          {/* @TODO(kerrya) put this back after pending state works properly */}
          {/* { pending && <LoadingSpinner />} */}
          End session
        </Button>
      )}
    </Session>
  );
};

export default OAuth2Session;
