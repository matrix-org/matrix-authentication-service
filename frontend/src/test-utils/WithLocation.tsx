// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// @vitest-environment happy-dom

import { Provider } from "jotai";
import { useHydrateAtoms } from "jotai/utils";

import { appConfigAtom, locationAtom } from "../routing";

const HydrateLocation: React.FC<React.PropsWithChildren<{ path: string }>> = ({
  children,
  path,
}) => {
  useHydrateAtoms([
    [appConfigAtom, { root: "/" }],
    [locationAtom, { pathname: path }],
  ]);
  return <>{children}</>;
};

/**
 * Utility for testing components that rely on routing or location
 * For example any component that includes a <Link />
 * Eg:
 * ```
 * const component = create(
      <WithLocation path="/">
        <NavItem route={{ type: "profile" }}>Active</NavItem>
      </WithLocation>,
    );
 * ```
 */
export const WithLocation: React.FC<
  React.PropsWithChildren<{ path?: string }>
> = ({ children, path }) => {
  return (
    <Provider>
      <HydrateLocation path={path || "/"}>{children}</HydrateLocation>
    </Provider>
  );
};
