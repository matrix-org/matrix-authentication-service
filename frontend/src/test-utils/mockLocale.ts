// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { vi } from "vitest";

/**
 * Mock the locale on Intl.DateTimeFormat
 * To achieve stable formatted dates across environments
 * Defaults to `en-GB`
 */
export const mockLocale = (defaultLocale = "en-GB"): void => {
  const { DateTimeFormat } = Intl;
  vi.spyOn(Intl, "DateTimeFormat").mockImplementation(
    (
      locales?: Intl.LocalesArgument,
      options?: Intl.DateTimeFormatOptions | undefined,
    ) => new DateTimeFormat(locales || defaultLocale, options),
  );
};
