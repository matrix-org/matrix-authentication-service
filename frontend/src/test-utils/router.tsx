// Copyright 2024 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {
  RouterProvider,
  createMemoryHistory,
  createRootRoute,
  createRoute,
  createRouter,
} from "@tanstack/react-router";

const rootRoute = createRootRoute();
const index = createRoute({ getParentRoute: () => rootRoute, path: "/" });
rootRoute.addChildren([index]);

const router = createRouter({
  history: createMemoryHistory(),
  routeTree: rootRoute,
});

const routerReady = router.load();
// Because we also use this in the stories, we need to make sure we call this only in tests
if (import.meta.vitest) {
  // Make sure the router is ready before running any tests
  import("vitest").then(({ beforeAll }) =>
    beforeAll(async () => await routerReady),
  );
}

export const DummyRouter: React.FC<React.PropsWithChildren> = ({
  children,
}) => (
  /** @ts-expect-error: The router we pass doesn't match the "real" router, which is fine for tests */
  <RouterProvider router={router} defaultComponent={() => children} />
);
