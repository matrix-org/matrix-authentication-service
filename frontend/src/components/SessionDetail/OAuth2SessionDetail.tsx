// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { useSetAtom } from "jotai";

import { FragmentType, useFragment } from "../../gql";
import { Link } from "../../routing";
import { getDeviceIdFromScope } from "../../utils/deviceIdFromScope";
import BlockList from "../BlockList/BlockList";
import DateTime from "../DateTime";
import { OAUTH2_SESSION_FRAGMENT, endSessionFamily } from "../OAuth2Session";
import ClientAvatar from "../Session/ClientAvatar";
import EndSessionButton from "../Session/EndSessionButton";

import SessionDetails from "./SessionDetails";
import SessionHeader from "./SessionHeader";

type Props = {
  session: FragmentType<typeof OAUTH2_SESSION_FRAGMENT>;
};

const OAuth2SessionDetail: React.FC<Props> = ({ session }) => {
  const data = useFragment(OAUTH2_SESSION_FRAGMENT, session);
  const endSession = useSetAtom(endSessionFamily(data.id));

  const onSessionEnd = async (): Promise<void> => {
    await endSession();
  };

  const deviceId = getDeviceIdFromScope(data.scope);

  const scopes = data.scope.split(" ");

  const finishedAt = data.finishedAt
    ? [{ label: "Finished", value: <DateTime datetime={data.createdAt} /> }]
    : [];

  const ipAddress = data.ipAddress
    ? [{ label: "IP Address", value: <code>{data.ipAddress}</code> }]
    : [];

  const sessionDetails = [
    { label: "ID", value: <code>{data.id}</code> },
    { label: "Device ID", value: <code>{deviceId}</code> },
    { label: "Signed in", value: <DateTime datetime={data.createdAt} /> },
    ...finishedAt,
    ...ipAddress,
    {
      label: "Scopes",
      value: (
        <div>
          {scopes.map((scope) => (
            <code key={scope}>{scope}</code>
          ))}
        </div>
      ),
    },
  ];

  const clientTitle = (
    <Link route={{ type: "client", id: data.client.id }}>Client</Link>
  );
  const clientDetails = [
    {
      label: "Name",
      value: (
        <>
          <ClientAvatar
            name={data.client.clientName || data.client.clientId}
            logoUri={data.client.logoUri || undefined}
            size="var(--cpd-space-4x)"
          />
          {data.client.clientName}
        </>
      ),
    },
    { label: "ID", value: <code>{data.client.clientId}</code> },
    {
      label: "Uri",
      value: (
        <a target="_blank" href={data.client.clientUri}>
          {data.client.clientUri}
        </a>
      ),
    },
  ];

  return (
    <BlockList>
      <SessionHeader
        backToRoute={{
          type: "sessions-overview",
        }}
      >
        {deviceId || data.id}
      </SessionHeader>
      <SessionDetails title="Session" details={sessionDetails} />
      <SessionDetails title={clientTitle} details={clientDetails} />
      {!data.finishedAt && <EndSessionButton endSession={onSessionEnd} />}
    </BlockList>
  );
};

export default OAuth2SessionDetail;
