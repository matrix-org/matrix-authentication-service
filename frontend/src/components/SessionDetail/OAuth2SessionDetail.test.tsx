// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// @vitest-environment happy-dom

import { render, cleanup } from "@testing-library/react";
import { describe, expect, it, afterEach } from "vitest";

import { makeFragmentData } from "../../gql/fragment-masking";
import { WithLocation } from "../../test-utils/WithLocation";
import { OAUTH2_SESSION_FRAGMENT } from "../OAuth2Session";

import OAuth2SessionDetail from "./OAuth2SessionDetail";

describe("<OAuth2SessionDetail>", () => {
  const baseSession = {
    id: "session-id",
    scope:
      "openid urn:matrix:org.matrix.msc2967.client:api:* urn:matrix:org.matrix.msc2967.client:device:abcd1234",
    createdAt: "2023-06-29T03:35:17.451292+00:00",
    client: {
      id: "test-id",
      clientId: "test-client-id",
      clientName: "Element",
      clientUri: "https://element.io",
    },
  };
  afterEach(cleanup);

  it("renders session details", () => {
    const data = makeFragmentData(baseSession, OAUTH2_SESSION_FRAGMENT);

    const { container } = render(
      <WithLocation>
        <OAuth2SessionDetail session={data} />
      </WithLocation>,
    );

    expect(container).toMatchSnapshot();
  });

  it("renders a finished session details", () => {
    const data = makeFragmentData(
      {
        ...baseSession,
        finishedAt: "2023-07-29T03:35:17.451292+00:00",
      },
      OAUTH2_SESSION_FRAGMENT,
    );

    const { getByText, queryByText } = render(
      <WithLocation>
        <OAuth2SessionDetail session={data} />
      </WithLocation>,
    );

    expect(getByText("Finished")).toBeTruthy();
    // no end session button
    expect(queryByText("End session")).toBeFalsy();
  });
});
