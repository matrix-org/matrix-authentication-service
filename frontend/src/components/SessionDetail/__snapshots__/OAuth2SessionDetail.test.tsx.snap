// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<OAuth2SessionDetail> > renders session details 1`] = `
<div>
  <div
    class="_blockList_f8cc7f"
  >
    <header
      class="_header_92353c"
    >
      <button
        class="_icon-button_1segd_17"
        style="--cpd-icon-button-size: 32px;"
        type="button"
      >
        <svg
          class="cpd-icon"
          fill="none"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.2071 5.29289C12.5976 5.68342 12.5976 6.31658 12.2071 6.70711L7.91421 11H18.5C19.0523 11 19.5 11.4477 19.5 12C19.5 12.5523 19.0523 13 18.5 13H7.91421L12.2071 17.2929C12.5976 17.6834 12.5976 18.3166 12.2071 18.7071C11.8166 19.0976 11.1834 19.0976 10.7929 18.7071L4.79289 12.7071C4.40237 12.3166 4.40237 11.6834 4.79289 11.2929L10.7929 5.29289C11.1834 4.90237 11.8166 4.90237 12.2071 5.29289Z"
            fill="currentColor"
          />
        </svg>
      </button>
      <h3
        class="_font-heading-md-semibold_1jx6b_121"
      >
        abcd1234
      </h3>
    </header>
    <div
      class="_block_17898c"
    >
      <h6
        class="_font-body-md-semibold_1jx6b_64"
      >
        Session
      </h6>
      <ul
        class="_list_040867"
      >
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            ID
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            <code>
              session-id
            </code>
          </p>
        </li>
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            Device ID
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            <code>
              abcd1234
            </code>
          </p>
        </li>
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            Signed in
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            <code>
              2023-06-29T03:35:17.451292+00:00
            </code>
          </p>
        </li>
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            Scopes
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            <p>
              <code>
                openid
              </code>
            </p>
            <p>
              <code>
                urn:matrix:org.matrix.msc2967.client:api:*
              </code>
            </p>
            <p>
              <code>
                urn:matrix:org.matrix.msc2967.client:device:abcd1234
              </code>
            </p>
          </p>
        </li>
      </ul>
    </div>
    <div
      class="_block_17898c"
    >
      <h6
        class="_font-body-md-semibold_1jx6b_64"
      >
        <a
          class=""
          href="/clients/test-id"
        >
          Client
        </a>
      </h6>
      <ul
        class="_list_040867"
      >
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            Name
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            Element
          </p>
        </li>
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            ID
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            <code>
              test-client-id
            </code>
          </p>
        </li>
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            Uri
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            <a
              href="https://element.io"
              target="_blank"
            >
              https://element.io
            </a>
          </p>
        </li>
      </ul>
    </div>
    <button
      class="_button_1raud_17"
      data-kind="destructive"
      data-size="sm"
      role="button"
      tabindex="0"
    >
      End session
    </button>
  </div>
</div>
`;
